src/dataflow/src/render/sources.rs

```diff
@@ -801,8 +814,41 @@ where
     K2: Codec + timely::Data,
     V2: Codec + timely::Data,
 {
-    let sealed_stream =
-        stream.conditional_seal(&source_name, bindings_stream, write, bindings_write);
+    let sealed_stream = stream.conditional_seal(
+        &source_name,
+        bindings_stream,
+        write.clone(),
+        bindings_write.clone(),
+    );
+
+    // The compaction since of persisted sources is upper-bounded by two frontiers: a) the
+    // compaction that the coordinator allows, and b) the seal frontier of the involved persistent
+    // collections.
+
+    // We need to obey a) because that's what the coordinator wants, and b) because if we persisted
+    // past (or up to) the seal frontier we would not be able to distinguish between valid and
+    // invalid updates based on the upper seal frontier when restarting.
+    //
+    // We get the upper seal frontier using a probe handle, and we expect the coordinator to call
+    // `allow_compaction` on the stashed `CompactionHandle` to present it's view of what should be
+    // allowed.
+
+    let mut probe: ProbeHandle<Timestamp> = ProbeHandle::new();
+    sealed_stream.probe_with(&mut probe);
+
+    let mut compaction_handle = MultiWriteHandle::new(&write);
```

Comment:
nit: this function previously was only doing dataflow rendering and now it does dataflow rendering _and_ managing persisted source state and handling tokens - can we maybe extract out the "wire up allow compaction bits" into another function?